src/test/clojure/clara/test_rules.clj

```diff
@@ -5025,3 +5025,129 @@
+                                             (= ?location location)
+                                             (> ?temperature windspeed)]])
+        temperature-query (assoc query-template :name "temperature-query")
+        temperature-fact (->Temperature nil "MCI")
+        windspeed-fact (->WindSpeed 30 "MCI")]
+    (assert-ex-data {:constraint-exp '(> ?temperature windspeed)
+                     :bindings {:?location "MCI", :?temperature nil}
+                     :fact windspeed-fact
+                     :queries ["temperature-query"]
+                     :productions []
+                     :condition [clara.rules.testfacts.WindSpeed
+                                 '(= ?location location)
+                                 '(> ?temperature windspeed)]}
+                    (-> (mk-session [temperature-query] :cache false)
+                        (insert temperature-fact)
+                        (insert windspeed-fact)
+                        (fire-rules)
+                        (query temperature-query)))))
+
+(deftest test-expression-join-query-constraint-exception-from-fact
+  (let [query-template (dsl/parse-query [] [[Temperature
+                                             (= ?location location)
+                                             (= ?temperature temperature)]
+                                            [WindSpeed
+                                             (= ?location location)
+                                             (> ?temperature windspeed)]])
+        temperature-query (assoc query-template :name "temperature-query")
+        temperature-fact (->Temperature 90 "MCI")
+        windspeed-fact (->WindSpeed nil "MCI")]
+    (assert-ex-data {:constraint-exp '(> ?temperature windspeed)
+                     :bindings {:?location "MCI", :?temperature 90}
+                     :fact windspeed-fact
+                     :queries ["temperature-query"]
+                     :productions []
+                     :condition [clara.rules.testfacts.WindSpeed
+                                 '(= ?location location)
+                                 '(> ?temperature windspeed)]}
+                    (-> (mk-session [temperature-query] :cache false)
+                        (insert temperature-fact)
+                        (insert windspeed-fact)
+                        (fire-rules)
+                        (query temperature-query)))))
+
+(deftest test-single-condition-accum-constraint-exception
+  (let [query-template (dsl/parse-query [] [[?wind <- (acc/all) :from [Temperature (> temperature 0)]]])
+        temperature-query (assoc query-template :name "temperature-query")
+        temperature-fact (->Temperature nil "MCI")]
+    (assert-ex-data {:constraint-exp '(> temperature 0)
+                     :bindings {}
+                     :fact temperature-fact
+                     :queries ["temperature-query"] ; ordering may change with implementation details
+                     :productions []}
```

Comment:
The condition structure was too brittle and I worried that it would easily in the future. Not a problem with the new protocols.